pubsub/gochannel/fanin.go

```diff
@@ -0,0 +1,139 @@
+package gochannel
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+// FanIn is a component that receives messages from 1..N topics from a subscriber and publishes them
+// on a specified topic. In effect, messages are "multiplexed".
+//
+// You need to call AddSubscription method with a slice of topics that you want to
+// listen to and a topic you want them to be published to.
+// This needs to be done *before* starting the FanIn.
+//
+// FanIn exposes the standard Subscriber interface.
+type FanIn struct {
+	internalPubSub *GoChannel
+	internalRouter *message.Router
+
+	subscriber message.Subscriber
+
+	logger watermill.LoggerAdapter
+
+	subscriptions  map[string]struct{}
+	subscribedLock sync.Mutex
+}
+
+// NewFanIn creates a new FanIn.
+func NewFanIn(
+	subscriber message.Subscriber,
+	logger watermill.LoggerAdapter,
+) (*FanIn, error) {
+	if subscriber == nil {
+		return nil, errors.New("missing subscriber")
+	}
+	if logger == nil {
+		logger = watermill.NopLogger{}
+	}
+
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &FanIn{
+		internalPubSub: NewGoChannel(Config{}, logger),
+		internalRouter: router,
+		logger:         logger,
+		subscriber:     subscriber,
+		subscriptions:  map[string]struct{}{},
+	}, nil
+}
+
+// AddSubscription adds an internal subscriptions.
+// You need to call this method with slice of `fromTopics` and `toTopic` before the FanIn is started.
+// AddSubscription is idempotent.
+func (f *FanIn) AddSubscription(fromTopics []string, toTopic string) {
```

Comment:
> You need to call this method with slice of `fromTopics` and `toTopic` before the FanIn is started.
so maybe it's better to add it to the constructor (and add some kind of config). In that case design would be simpler :wink: